pandas/core/frame.py

```diff
@@ -7834,7 +7834,10 @@ def corr(self, method="pearson", min_periods=1) -> "DataFrame":
         numeric_df = self._get_numeric_data()
         cols = numeric_df.columns
         idx = cols.copy()
-        mat = numeric_df.values
+        mat = numeric_df.to_numpy()
+        if is_object_dtype(mat.dtype):
+            # We end up with an object array if pd.NA is present
+            mat[isna(mat)] = np.nan
```

Comment:
This feels like a hack, and it seems it should somehow be possible to cast this whole DataFrame to float using to_numpy but that doesn't seem to work

Reply:
pass `na_value=np.nan`

Reply:
I was trying that originally but it turns out DataFrame.to_numpy doesn't have an na_value argument (seems only Series and EAs do)

Reply:
hmm @TomAugspurger

Reply:
https://github.com/pandas-dev/pandas/issues/33820

Reply:
We'll want to avoid going through object dtype at all. Can you specify `dtype="float"`? Or does that break other things?

Reply:
That raises an error:
```python
TypeError: float() argument must be a string or a number, not 'NAType'
```

Reply:
That matches the behavior of Series.to_numpy(). We need to have the NA value passed through as `np.nan` I think.

Reply:
I think as a workaround astyping to float then calling to_numpy works, then later could use na_value for DataFrame.to_numpy directly?